### ISSUE TITLE
[Epic] Multi-process timely clusters

### ISSUE THREAD

**Issue by frankmcsherry** [MaterializeInc/materialize#11238]
Timely dataflow and differential dataflow already support multi-process installations and we want to expose this functionality to customers.

### Goals
* Understand scaling behavior with multi-process clusters
* Productionize existing multi-process code

### Non-Goals
* Auto-scaling clusters
* Support for sinks (blocked by https://github.com/MaterializeInc/materialize/issues/11501)
* Sizing of cluster instance(s) (part of #11240)

### Acceptance criteria

Users can create horizontally-scaled cluster instances. Horizontally-scaled instances offer soft-scaling, which enables processing larger workloads at similar performance levels.

### Subtasks

- [ ] Establish base scaling characteristics 
  - [x] https://github.com/MaterializeInc/materialize/pull/11791
  - [ ] https://github.com/MaterializeInc/materialize/issues/11947
  - [ ] https://github.com/MaterializeInc/materialize/issues/12069
- [x] https://github.com/MaterializeInc/materialize/issues/11810 (shared with #11240)
- [x] (The "orchestrator changes" and "cloud changes" parts of https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20220413_cluster_replica.md also still need to be done)
- [x] https://github.com/MaterializeInc/materialize/issues/12053
- [x] https://github.com/MaterializeInc/materialize/issues/11710 (for debugging)
- [ ] https://github.com/MaterializeInc/materialize/issues/12816
- [ ] **Bugs**
  - [ ] https://github.com/MaterializeInc/materialize/issues/12233

### Testing tasks
- [ ] Make sure multi-process clusters are represented in the feature benchmark
- [ ] Run testdrive Nightly against a multi-process cluster
- [ ] Allow Zippy to create and tear-down multi-process clusters as the test is running 
- [ ] Make sure multi-process clusters are represented in upgrade and other failure- and restart-containing testing frameworks

### Related

* [design/20220214_create_cluster.md](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20220214_create_cluster.md)
* [design/20220413_cluster_replica.md](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20220413_cluster_replica.md)
  * There is some overlap with #11240. Coordinate in order to avoid duplicated work and unnecessary conflicts.

### Open Questions

* Are there more places that assume unique processes?
* Network reliability?
    * See also: https://github.com/TimelyDataflow/timely-dataflow/issues/386
* Fault tolerance?
* Is there a work-around for sinks?
* Are the memory issues from https://github.com/MaterializeInc/materialize/issues/11947 important enough to be part of M1 ?
